src/rdkafka_sasl_oauthbearer_oidc.c

```diff
@@ -0,0 +1,397 @@
+
+        *headers = curl_slist_append(*headers, accept_header);
+        *headers = curl_slist_append(*headers, authorization_base64_header);
+        rd_free(authorization_base64_header);
+        return rd_true;
+}
+
+
+/**
+ * @brief Implementation of Oauth/OIDC token refresh call back function,
+ *        will receive the json response after HTTP call from token provider,
+ *        then extract the jwt from the json response, and forward it to
+ *        the broker.
+ */
+void rd_kafka_conf_set_oauthbearer_oidc_token_refresh_cb (rd_kafka_t *rk,
+        const char *oauthbearer_config, void *opaque) {
+        static const char *data = "grant_type=client_credentials,scope=%s";
+        static const char *token_key = "access_token";
+        static const long timeout = 20;
+        static const size_t retry = 3;
+
+        cJSON *json = NULL;
+        rd_http_error_t *herr;
+        char *jwt_token = NULL;
+        struct curl_slist *headers = NULL;
+        char *data_to_token;
+        size_t data_to_token_size;
+        const char *scope;
+        const char *token_url;
+        rd_bool_t build_headers_status;
+        char *error;
+
+        token_url = rk->rk_conf.sasl.oauthbearer.token_endpoint_url;
+        if (token_url == NULL) {
+                rd_kafka_log(rk, LOG_ERR, "OAUTH/OIDC",
+                             "token_url is a required field");
+                return;
+        }
+
+        build_headers_status = rd_false;
+        build_headers_status = build_headers(&rk->rk_conf, &headers, &error);
+        if (!build_headers_status) {
+                rd_kafka_log(rk, LOG_ERR, "OAUTH/OIDC",
+                             "Failed to build headers: %s",
+                             error);
+                rd_free(error);
+                return;
+        }
+
+        /* Build data field */
+        scope = rk->rk_conf.sasl.oauthbearer.scope;
+        if (scope == NULL) {
+                rd_kafka_log(rk, LOG_ERR, "OAUTH/OIDC",
+                             "scope is a required field");
+                return;
+        }
+        data_to_token_size = strlen(data) + strlen(scope) - 1;
+        data_to_token = rd_malloc(data_to_token_size);
+        rd_snprintf(data_to_token, data_to_token_size, data, scope);
+        herr = rd_http_post_json(token_url,
+                                 &headers,
+                                 data_to_token,
+                                 data_to_token_size,
+                                 timeout,
+                                 retry,
+                                 &json);
+        if (unlikely(herr != NULL)) {
+                rd_kafka_log(rk, LOG_ERR, "OAUTH/OIDC",
+                             "Failed to receive json result from HTTP(S), "
+                             "returned error code: %d, returned error string: "
+                             "%s", herr->code, herr->errstr);
+                rd_http_error_destroy(herr);
+                goto exit;
```

Comment:
`exit` may give the reader the idea that `exit()` will be called eventually. We typically use `done` for these types of gotos.